kernel/arch/dreamcast/hardware/cdrom.c

```diff
@@ -408,35 +408,66 @@ int cdrom_spin_down(void) {
     return rv;
 }
 
+/*
+    Unlocks G1 ATA DMA access to all memory on the root bus, not just system memory.
+    Patches syscall region where the DMA protection register is set,
+    ensuring it allows broader memory access, and updates the register accordingly.
+ */
+static void unlock_dma_memory(void) {
+    uint32_t i, patched = 0;
+    volatile uint32_t *prot_reg = (uint32_t *)(G1_ATA_DMA_PROTECTION | MEM_AREA_P2_BASE);
+    uintptr_t patch_addr[2] = {0x0c001c20, 0x0c0023fc};
+
+    for(i = 0; i < sizeof(patch_addr) / sizeof(uintptr_t); ++i) {
+        if(*(uint32_t *)(patch_addr[i] | MEM_AREA_P2_BASE) == (uint32_t)G1_ATA_DMA_UNLOCK_SYSMEM) {
+            *(uint32_t *)(patch_addr[i] | MEM_AREA_P2_BASE) = G1_ATA_DMA_UNLOCK_ALLMEM;
+            ++patched;
+        }
+    }
+    if(patched) {
+        icache_flush_range(patch_addr[0] | MEM_AREA_P1_BASE, patch_addr[1] - patch_addr[0]);
```

Comment:
This won't flush the `patch_addr[1]` address, as it will stop just before - is that a problem? Maybe you should add a +4 there?